File: python/src/skills_utils/repo_parse.py
from __future__ import annotations

import re

_GITHUB_URL = re.compile(
    r"(?:https?://)?(?:www\.)?github\.com/([^/]+)/([^/#?]+)",
    re.IGNORECASE,
)


def parse_github_repo(owner: str, repo: str) -> tuple[str, str]:
    """Normalize owner/repo from plain names, owner/repo strings, or GitHub URLs."""
    owner = (owner or "").strip()
    repo = (repo or "").strip()

    for candidate in (repo, owner, f"{owner}/{repo}".strip("/")):
        if not candidate:
            continue
        match = _GITHUB_URL.search(candidate)
        if match:
            return match.group(1), match.group(2).removesuffix(".git")

    if "/" in owner and not repo:
        parts = owner.strip("/").split("/", 1)
        return parts[0], parts[1].removesuffix(".git")

    if "/" in repo and not owner:
        parts = repo.strip("/").split("/", 1)
        return parts[0], parts[1].removesuffix(".git")

    return owner, repo.removesuffix(".git")


def default_repo_slug(owner: str, repo: str) -> str:
    """Short local namespace; prefer repo name unless it is too generic."""
    if repo in {"skills", "agent-skills"}:
        return owner
    return repo

File: node/src/repoParse.ts
const GITHUB_URL = /(?:https?:\/\/)?(?:www\.)?github\.com\/([^/]+)\/([^/#?]+)/i;

export function parseGithubRepo(owner: string, repo: string): [string, string] {
  const o = (owner ?? "").trim();
  const r = (repo ?? "").trim();
  for (const candidate of [r, o, `${o}/${r}`.replace(/\/+$/, "")]) {
    if (!candidate) continue;
    const match = candidate.match(GITHUB_URL);
    if (match) return [match[1], match[2].replace(/\.git$/, "")];
  }
  if (o.includes("/") && !r) {
    const [a, b] = o.split("/", 2);
    return [a, b.replace(/\.git$/, "")];
  }
  if (r.includes("/") && !o) {
    const [a, b] = r.split("/", 2);
    return [a, b.replace(/\.git$/, "")];
  }
  return [o, r.replace(/\.git$/, "")];
}

export function defaultRepoSlug(owner: string, repo: string): string {
  if (repo === "skills" || repo === "agent-skills") return owner;
  return repo;
}
